File: algorithms/C/Count Nodes With the Highest Score/Count Nodes With the Highest Score.java
class Solution {
    long max = 0, res = 0;
    public int countHighestScoreNodes(int[] parents) {
        Map<Integer, List<Integer>> hm = new HashMap();
        for(int i = 0; i < parents.length; i++) {   // build the tree
            hm.computeIfAbsent(parents[i], x ->new ArrayList<>()).add(i);
        }
        dfs(0, parents.length, hm); // traverse the tree to get the result
        return (int)res;
    }
    int dfs(int s, int n, Map<Integer, List<Integer>> hm) {
        int sum = 1;
        long mult = 1L;
        for(int child : hm.getOrDefault(s, new ArrayList<>())) {
            int count = dfs(child, n, hm); // subtree node count
            sum += count;
            mult *= count; // multiply the result by children size
        }     
        mult *= (s == 0 ? 1L : n - sum); // multiply the result by remain size except self and children size(the nodes through parent)
        if(mult > max) {
            max = mult;
            res = 1;
        } else if (mult == max) {
            res++;
        }
        return sum; // return the node count of the tree rooted at s
    }
}

File: scripts/algorithms/C/Count Nodes With the Highest Score/Count Nodes With the Highest Score.py
class Solution:
    def countHighestScoreNodes(self, parents: List[int]) -> int:
        graph = collections.defaultdict(list)
        for node, parent in enumerate(parents):  # build graph
            graph[parent].append(node)
        n = len(parents)                         # total number of nodes
        d = collections.Counter()
        def count_nodes(node):                   # number of children node + self
            p, s = 1, 0                          # p: product, s: sum
            for child in graph[node]:            # for each child (only 2 at maximum)
                res = count_nodes(child)         # get its nodes count
                p *= res                         # take the product
                s += res                         # take the sum
            p *= max(1, n - 1 - s)               # times up-branch (number of nodes other than left, right children ans itself)
            d[p] += 1                            # count the product
            return s + 1                         # return number of children node + 1 (self)
        count_nodes(0)                           # starting from root (0)
        return d[max(d.keys())]                  # return max count
